File: src/main/java/mizuki/audience/RedisManager.java
package mizuki.audience;

import org.apache.commons.pool2.impl.GenericObjectPoolConfig;
import redis.clients.jedis.Jedis;
import redis.clients.jedis.JedisPool;
import redis.clients.jedis.exceptions.JedisException;

class RedisManager {
//    private final int REDIS_INDEX = Integer.parseInt(System.getenv("REDIS_INDEX"));
//    private final String REDIS_HOST = System.getenv("REDIS_HOST");
    private JedisPool jedisPool;
    private static final RedisManager redisManager = new RedisManager();

    private RedisManager() {
        GenericObjectPoolConfig JedisPoolConfig = new GenericObjectPoolConfig();
        JedisPoolConfig.setMaxIdle(200);
        JedisPoolConfig.setMinIdle(20);
        jedisPool = new JedisPool(JedisPoolConfig, "localhost");
    }

    static RedisManager getInstance() {
        return redisManager;
    }

    Jedis getJedisResource() {
        try {
            if (this.jedisPool.getResource().isConnected()) {
                return jedisPool.getResource();
            }
        } catch (JedisException e) {
            // @todo re-connect refactoring
            e.printStackTrace();
            GenericObjectPoolConfig JedisPoolConfig = new GenericObjectPoolConfig();
            JedisPoolConfig.setMaxIdle(200);
            JedisPoolConfig.setMinIdle(20);
            this.jedisPool = new JedisPool(JedisPoolConfig, System.getenv("REDIS_HOST"));
        } finally {
            return this.jedisPool.getResource();
        }
    }
}

File: src/main/java/mizuki/audience/RedisManager.kt
package mizuki.audience

import org.apache.commons.pool2.impl.GenericObjectPoolConfig
import redis.clients.jedis.Jedis
import redis.clients.jedis.JedisPool
import redis.clients.jedis.exceptions.JedisException



object RedisManager {
    private var jedisPool: JedisPool = this.connect()
    init {
        this.connect()
    }
    fun getJedisResource(): Jedis {
        try {
            if (this.jedisPool.resource.isConnected) {
                this.jedisPool = this.connect()
            }
        } catch (e: JedisException) {
            e.printStackTrace()
        }

        return this.jedisPool.resource
    }

    private fun connect() : JedisPool {
        val JedisPoolConfig = GenericObjectPoolConfig()
        JedisPoolConfig.maxIdle = 200
        JedisPoolConfig.minIdle = 20
        return JedisPool(JedisPoolConfig, "localhost")
    }
}
